iocore/net/QUICNetVConnection.cc

```diff
@@ -503,24 +542,33 @@ QUICNetVConnection::state_handshake(int event, Event *data)
 {
   SCOPED_MUTEX_LOCK(lock, this->mutex, this_ethread());
 
-  if (this->_handshake_handler && this->_handshake_handler->is_completed()) {
-    this->_switch_to_established_state();
-    return this->handleEvent(event, data);
-  }
-
   QUICErrorUPtr error = QUICErrorUPtr(new QUICNoError());
 
   switch (event) {
   case QUIC_EVENT_PACKET_READ_READY: {
     QUICPacketCreationResult result;
-    QUICPacketUPtr packet = this->_dequeue_recv_packet(result);
-    if (result == QUICPacketCreationResult::NOT_READY) {
-      error = QUICErrorUPtr(new QUICNoError());
-    } else if (result == QUICPacketCreationResult::FAILED) {
-      error = QUICConnectionErrorUPtr(new QUICConnectionError(QUICTransErrorCode::TLS_FATAL_ALERT_GENERATED));
-    } else {
-      error = this->_state_handshake_process_packet(std::move(packet));
+    CountQueue<UDPPacket> remainder;
+    do {
+      QUICPacketUPtr packet = this->_dequeue_recv_packet(remainder, result);
+      if (result == QUICPacketCreationResult::NOT_READY) {
+        error = QUICErrorUPtr(new QUICNoError());
+      } else if (result == QUICPacketCreationResult::FAILED) {
+        error = QUICConnectionErrorUPtr(new QUICConnectionError(QUICTransErrorCode::TLS_FATAL_ALERT_GENERATED));
+      } else {
+        error = this->_state_handshake_process_packet(std::move(packet));
+      }
+
+      if (this->_handshake_handler && this->_handshake_handler->is_completed()) {
+        this->_switch_to_established_state();
+        return this->handleEvent(event, data);
+      }
+    } while (result == QUICPacketCreationResult::SUCCESS && error->cls == QUICErrorClass::NONE);
+
+    if (remainder.size > 0) {
+      this->_packet_recv_queue.append(remainder);
```

Comment:
I agree we should schedule the event here (not inside the loop), however, may be we should requeue packets inside `_dequeue_recv_packet ` instead of using `reminder`. If packets A to C are enqueued like A(Key1), C(Key2), and B(Key1), we can process all packets actually.

Also, reordering packets is not required on the spec, as far as I know. So may be we can just discard unordered packets. What do you think?

Reply:
The reason I don't requeue pakcet inside _dequeue_recv_packet since we only dequeue one udp packet and generate one quic packet in  `_dequeue_recv_packet`, and get another packet in the next loop . If we requeue the (not_ready) packet, we may get a bad loop.

Reply:
But if client don't send the B any more. the packet is never be freed.

So I agree with you discard may be a good choice.

Reply:
> If we requeue the (not_ready) packet, we may get a bad loop.

Yeah, we would probably need a loop counter if we requeue packets so we can break out the loop when there's no progress.

> But if client don't send the B any more. the packet is never be freed.

This reminded me that packets encrypted with unexpected key (e.g. unsupported versions) can keep in queue. It happened in the last interop.

Let's discard the packets.